src/platform/riot/coap_udp_ip_channel.c

```diff
@@ -0,0 +1,348 @@
+  coap_hdr_set_type(pdu.hdr, msg_type);
+
+  coap_opt_add_format(&pdu, COAP_FORMAT_TEXT);
+  ssize_t len = coap_opt_finish(&pdu, COAP_OPT_FINISH_PAYLOAD);
+
+  // Serialize message into CoAP payload buffer
+  int payload_len = serialize_to_protobuf(message, pdu.payload, pdu.payload_len);
+
+  if (payload_len < 0) {
+    LF_ERR(NET, "CoapUdpIpChannel: Could not encode protobuf");
+    return LF_ERR;
+  }
+
+  if (pdu.payload_len < payload_len) {
+    LF_ERR(NET, "CoapUdpIpChannel: Send CoAP message: msg buffer too small (%d < %d)", pdu.payload_len, payload_len);
+    return false;
+  }
+
+  // Update CoAP packet length based on serialized payload length
+  len = len + payload_len;
+
+  ssize_t bytes_sent = gcoap_req_send(self->write_buffer, len, remote, NULL, resp_handler, NULL, GCOAP_SOCKET_TYPE_UDP);
+  LF_DEBUG(NET, "CoapUdpIpChannel: Sending %d bytes", bytes_sent);
+  if (bytes_sent > 0) {
+    LF_DEBUG(NET, "CoapUdpIpChannel: Successfully sent");
+    return true;
+  }
+
+  return false;
+}
+
+static ssize_t _server_connect_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, coap_request_ctx_t *ctx) {
+  LF_DEBUG(NET, "CoapUdpIpChannel: Server connect handler");
+  CoapUdpIpChannel *self = _get_coap_channel_by_remote(ctx->remote);
+
+  // Error => return 401 (unauthorized)
+  if (self == NULL) {
+    LF_ERR(NET, "CoapUdpIpChannel: Server connect handler: Client has unknown IP address");
+    return gcoap_response(pdu, buf, len, COAP_CODE_UNAUTHORIZED);
+  }
+
+  // Error => return 503 (service unavailable)
+  if (self->state == NETWORK_CHANNEL_STATE_CLOSED) {
+    LF_ERR(NET, "CoapUdpIpChannel: Server connect handler: Channel is closed");
+    return gcoap_response(pdu, buf, len, COAP_CODE_SERVICE_UNAVAILABLE);
+  }
+
+  // Do not update the state here.
+  // The connected state is only determined by if the client succeeds
+  // to connect to the server of the other federate.
+  // This way it is guaranteed that the servers and clients on both federates
+  // work correctly (at least for connecting).
+
+  // Success => return 204 (no content)
+  return gcoap_response(pdu, buf, len, COAP_CODE_204);
+}
+
+static ssize_t _server_disconnect_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, coap_request_ctx_t *ctx) {
+  LF_DEBUG(NET, "CoapUdpIpChannel: Server disconnect handler");
+  CoapUdpIpChannel *self = _get_coap_channel_by_remote(ctx->remote);
+
+  // Error => return 401 (unauthorized)
+  if (self == NULL) {
+    LF_ERR(NET, "CoapUdpIpChannel: Server disconnect handler: Client has unknown IP address");
+    return gcoap_response(pdu, buf, len, COAP_CODE_UNAUTHORIZED);
+  }
+
+  // Update state because it does not make sense to send data to a closed connection.
+  // Only set state to disconnected if not already in more restrictive state
+  if (self->state != NETWORK_CHANNEL_STATE_UNINITIALIZED && self->state != NETWORK_CHANNEL_STATE_CLOSED) {
+    self->state = NETWORK_CHANNEL_STATE_CLOSED;
```

Comment:
Here I think we must call `network_channel_state_updated` so that the runtime knows about it?

Reply:
I think that function also might need a critical section to avoid race condition

Reply:
>Here I think we must call network_channel_state_updated so that the runtime knows about it?

Should we maybe always call this callback whenever we change state?

Reply:
Perhaps, I think this whole interaction (runtime <-> network channels) needs some careful thought. What we should avoid is any ambiguity in how information is communicated between them. Currently, for the API calls that deal with setting up the configuration: `open_connection`, `try_connect`, `try_reconnect`, `close_connection`, the state change is implied by the return code. If `LF_IN_PROGRESS` is returned, then the state should be `NETWORK_CHANNEL_STATE_CONNECTION_IN_PROGRESS`. So it is already quite bad that we also have this `network_channel_state_updated` callback which is called sometimes when the state is changed.

Maybe it is better to totally remove that callback, just modify the `state` variable. Then we have to always check the network channel state before we do any blocking wait on input. 

What do you think? I am worried that the interactions with the network channel will be the most complicated problem for our runtime, and that we must keep the interfaces very clear

Reply:
Yes I also found it a little bit duplicated to have these `LF_IN_PROGRESS` return codes and also a state with similar values. But we do need the state and it is useful to have `internally`.

I think because of the confusion it creates maybe ´NOT` expose it to outside of the channel and keep the state internal.

Instead of checking from the runtime each time we should maybe check inside of the channel functions if we can do the operation based on our internal state and if not return an appropriate `LF_ERR` code?

This way the runtime just gets an error code non-blocking of course and can continue just like it checks it itself, but the API is more clear and not duplicated between LF return codes and internal state.

What do you think?

Reply:
@erlingrj Sorry for the ping, just want to make sure, this is not lost in the amount of commits I pushed today while testing the CI 😳

Reply:
PS: I will think about the race condition tomorrow ;)

Reply:
Hm, the problem is that the "blocking" is actually not performed in the network channel (sorry I was not very accurate). Currently in `Scheduler_federated_acquire_tag` we check the `last_known_tag` of each input port and possibly go to interruptible sleep waiting for the time to elapse. 

In this function, or someplace before, we must check the state of the connection to detect whether the connection has been closed since last time we checked. We must be weary of a race condition here also, this function is executed in a critical section, but we must take care to call `new_async_event` on `Platform` if the connection state changes asynchronously (e.g. receive thread returns with EOF.

If you are interested in proposing something better here, I would gladly review it. We should probably merge this PR and handle this in a separate PR

Reply:
I think maybe we should discuss this in the next meeting on Thursday. I am not sure if I follow you correctly here

Reply:
Yes, lets discuss it then!